src/crab/wto.hpp

```diff
@@ -1,467 +1,180 @@
- * ENDORSEMENT BY GOVERNMENT AGENCY OR ANY PRIOR RECIPIENT OF ANY RESULTS,
- * RESULTING DESIGNS, HARDWARE, SOFTWARE PRODUCTS OR ANY OTHER APPLICATIONS
- * RESULTING FROM USE OF THE SUBJECT SOFTWARE.  FURTHER, GOVERNMENT AGENCY
- * DISCLAIMS ALL WARRANTIES AND LIABILITIES REGARDING THIRD-PARTY SOFTWARE,
- * IF PRESENT IN THE ORIGINAL SOFTWARE, AND DISTRIBUTES IT "AS IS."
- *
- * Waiver and Indemnity:  RECIPIENT AGREES TO WAIVE ANY AND ALL CLAIMS AGAINST
- * THE UNITED STATES GOVERNMENT, ITS CONTRACTORS AND SUBCONTRACTORS, AS WELL
- * AS ANY PRIOR RECIPIENT.  IF RECIPIENT'S USE OF THE SUBJECT SOFTWARE RESULTS
- * IN ANY LIABILITIES, DEMANDS, DAMAGES, EXPENSES OR LOSSES ARISING FROM SUCH
- * USE, INCLUDING ANY DAMAGES FROM PRODUCTS BASED ON, OR RESULTING FROM,
- * RECIPIENT'S USE OF THE SUBJECT SOFTWARE, RECIPIENT SHALL INDEMNIFY AND HOLD
- * HARMLESS THE UNITED STATES GOVERNMENT, ITS CONTRACTORS AND SUBCONTRACTORS,
- * AS WELL AS ANY PRIOR RECIPIENT, TO THE EXTENT PERMITTED BY LAW.
- * RECIPIENT'S SOLE REMEDY FOR ANY SUCH MATTER SHALL BE THE IMMEDIATE,
- * UNILATERAL TERMINATION OF THIS AGREEMENT.
- *
- ******************************************************************************/
-
+// Copyright (c) Prevail Verifier contributors.
+// SPDX-License-Identifier: MIT
 #pragma once
 
-#include <memory>
-#include <set>
+// This file implements Weak Topological Ordering as defined in
+// Bournacle, "Efficient chaotic iteration strategies with widenings", 1993
+// http://citeseerx.ist.psu.edu/viewdoc/summary?doi=10.1.1.38.3574
+//
+// Using the example from section 3.1 in the paper, the graph:
+//
+//         4 --> 5 <-> 6
+//         ^ \________ |
+//         |          vv
+//         3 <-------- 7
+//         ^           |
+//         |           v
+//   1 --> 2 --------> 8
+//
+// results in the WTO: 1 2 (3 4 (5 6) 7) 8
+// where a single vertex is represented via wto_vertex_t, and a
+// cycle such as (5 6) is represented via a wto_cycle_t.
+// Each arrow points to a wto_component_t, which can be either a
+// single vertex such as 8, or a cycle such as (5 6).
+
+#include <stack>
 #include <vector>
-#include <forward_list>
-#include <map>
-#include <variant>
-
-#include "crab/cfg.hpp"
-#include "crab/cfg_bgl.hpp"
-#include "crab/interval.hpp"
-#include "crab_utils/debug.hpp"
-
-namespace crab {
-
-
-using vertex_descriptor_t = typename boost::graph_traits<cfg_t>::vertex_descriptor;
-
-
-using out_edge_iterator_t = typename boost::graph_traits<cfg_t>::out_edge_iterator;
-
-
-using edge_descriptor_t = typename boost::graph_traits<cfg_t>::edge_descriptor;
-
-
-class wto_t;
-class wto_vertex_t;
-class wto_cycle_t;
-
-
-class wto_nesting_t final {
 
-    friend class wto_t;
-    friend class wto_vertex_t;
-    friend class wto_cycle_t;
+using label_t = crab::label_t;
+using wto_component_t = std::variant<std::shared_ptr<class wto_cycle_t>, label_t>;
+using wto_partition_t = std::vector<std::shared_ptr<wto_component_t>>;
```

Comment:
I tried making it use unique_ptr instead of shared_ptr but could never get it to compile (I'm not an expert on shared_ptr vs unique_ptr by any stretch) but the compiler issues were around trying to emplace_back into a vector.  I'm hoping it can be merged using shared_ptr and if someone else who's more a C++ expert than me wants to try to update it, that's fine :)

Reply:
I was mainly wondering if I understood the lifetime right. I think std::move should work for emplacing_back, but I don't always succeed with unique pointers either. I'll try that some time.